src/passes/OptimizeInstructions.cpp

```diff
@@ -3567,9 +3574,70 @@ struct OptimizeInstructions
         return left;
       }
     }
+    {
+      // (unsigned)  x + C1 >  C2   ==>  x >  C2-C1   if no overflowing
+      // (unsigned)  x + C1 >= C2   ==>  x >= C2-C1   if no overflowing
+      Binary* add;
+      Const* c1;
+      Const* c2;
+      if ((matches(
+             curr,
+             binary(GtU, binary(&add, Add, any(), ival(&c1)), ival(&c2))) ||
+           matches(
+             curr,
+             binary(GeU, binary(&add, Add, any(), ival(&c1)), ival(&c2)))) &&
+          !canOverflow(add)) {
+        if (c1->value.gtU(c2->value).getInteger()) {
+          // C2-C1 overflows. This is a situation that looks like this:
+          //   (unsigned)  x + 10 > 5
+          // The result is always true.
+          c1->value = Literal(int32_t(1));
+          c1->type = Type::i32;
+          return getDroppedChildrenAndAppend(curr, c1);
+        }
+        c2->value = c2->value.sub(c1->value);
+        curr->left = add->left;
+        return curr;
+      }
+    }
     return nullptr;
   }
 
+  // Returns true if the given binary operation can overflow. If we can't be
+  // sure either way, we return true, assuming the worst.
+  bool canOverflow(Binary* binary) {
+    using namespace Abstract;
+
+    // If we know nothing about a limit on the amount of bits on either side,
+    // give up.
+    auto typeMaxBits = getBitsForType(binary->type);
+    auto leftMaxBits = Bits::getMaxBits(binary->left, this);
+    if (leftMaxBits == typeMaxBits) {
+      return true;
+    }
+    auto rightMaxBits = Bits::getMaxBits(binary->right, this);
+    if (rightMaxBits == typeMaxBits) {
+      return true;
+    }
```

Comment:
What about use more generic approach?
```ts
if (max(getMaxBits(left), getMaxBits(right)) == typeMaxBits) {
  return true;
}
```

Reply:
I guess we don't use leftMaxBits/rightMaxBits in the code right now, but we do use them in a comment lower down, so I think it's clearer to save them to locals explicitly. But maybe using max would still be slightly shorter, good idea.